compiler/rustc_mir_transform/src/check_alignment.rs

```diff
@@ -63,69 +60,56 @@ impl<'tcx> MirPass<'tcx> for CheckAlignment {
     }
 }
 
-impl<'tcx, 'a> PointerFinder<'tcx, 'a> {
-    fn find_pointers(&mut self, statement: &Statement<'tcx>) -> Vec<(Place<'tcx>, Ty<'tcx>)> {
-        self.pointers.clear();
-        self.visit_statement(statement, Location::START);
-        core::mem::take(&mut self.pointers)
-    }
-}
-
 struct PointerFinder<'tcx, 'a> {
-    local_decls: &'a mut LocalDecls<'tcx>,
     tcx: TyCtxt<'tcx>,
-    def_id: DefId,
+    local_decls: &'a mut LocalDecls<'tcx>,
+    param_env: ParamEnv<'tcx>,
     pointers: Vec<(Place<'tcx>, Ty<'tcx>)>,
 }
 
 impl<'tcx, 'a> Visitor<'tcx> for PointerFinder<'tcx, 'a> {
     fn visit_rvalue(&mut self, rvalue: &Rvalue<'tcx>, location: Location) {
         if let Rvalue::AddressOf(..) = rvalue {
-            // Ignore dereferences inside of an AddressOf
+            // One of the primary uses of AddressOf is creating pointers to unaligned places, such
+            // as fields of packed structs.
             return;
         }
         self.super_rvalue(rvalue, location);
     }
 
-    fn visit_place(&mut self, place: &Place<'tcx>, context: PlaceContext, _location: Location) {
-        if let PlaceContext::NonUse(_) = context {
-            return;
-        }
+    fn visit_place(&mut self, place: &Place<'tcx>, context: PlaceContext, location: Location) {
         if !place.is_indirect() {
             return;
         }
```

Comment:
Instead of using `visit_rvalue`, what would it look like using a match over `context` to decide what to do?

Reply:
Good question. I think that approach probably works? `NonMutatingUseContext::Inspect` is a bit confusing to me, because it covers a lot of cases which I think are all partial reads of the Place? But then there's also `FakeRead` which documents

> When executed at runtime this is a nop.

So I feel like I should avoid checking it, but also

> Disallowed after drop elaboration.

So I feel like we won't encounter it and will be fine?

`SetDiscriminant` is also confusing because I think we'd want to check alignment for the write to the discriminant is valid, but checking Places with that context would assert alignment of the whole Place.

Reply:
Yeah, this check will complain for some cases of `SetDiscriminant(*ptr)` that Miri would allow (if the discriminant has lower alignment than the surrounding enum).

Also I'd just exclude `Inspect` then if it seems confusing. Better safe than sorry, we have tests ensuring the right things are covered.